src/x/serialize/decoder_fast.go

```diff
@@ -31,48 +31,71 @@ func TagValueFromEncodedTagsFast(
 	encodedTags []byte,
 	tagName []byte,
 ) ([]byte, bool, error) {
+	var (
+		length int
+		err    error
+	)
+	encodedTags, length, err = DecodeHeader(encodedTags)
+	if err != nil {
+		return nil, false, err
+	}
+
+	for i := 0; i < length; i++ {
+		var bytesName, bytesValue []byte
+
+		encodedTags, bytesName, bytesValue, err = DecodeTag(encodedTags)
+		if err != nil {
+			return nil, false, err
+		}
+
+		if bytes.Equal(bytesName, tagName) {
+			return bytesValue, true, nil
+		}
+	}
+
+	return nil, false, nil
+}
+
+func DecodeHeader(encodedTags []byte) ([]byte, int, error) {
 	total := len(encodedTags)
 	if total < 4 {
-		return nil, false, fmt.Errorf(
+		return nil, 0, fmt.Errorf(
 			"encoded tags too short: size=%d, need=%d", total, 4)
 	}
 
 	header := byteOrder.Uint16(encodedTags[:2])
 	encodedTags = encodedTags[2:]
 	if header != headerMagicNumber {
-		return nil, false, errIncorrectHeader
+		return nil, 0, errIncorrectHeader
 	}
 
 	length := int(byteOrder.Uint16(encodedTags[:2]))
 	encodedTags = encodedTags[2:]
+	return encodedTags, length, nil
+}
 
-	for i := 0; i < length; i++ {
-		if len(encodedTags) < 2 {
-			return nil, false, fmt.Errorf("missing size for tag name: index=%d", i)
-		}
-		numBytesName := int(byteOrder.Uint16(encodedTags[:2]))
-		if numBytesName <= 0 {
-			return nil, false, errEmptyTagNameLiteral
-		}
-		encodedTags = encodedTags[2:]
-
-		bytesName := encodedTags[:numBytesName]
-		encodedTags = encodedTags[numBytesName:]
+func DecodeTag(encodedTags []byte) ([]byte, []byte, []byte, error) {
+	if len(encodedTags) < 2 {
+		return nil, nil, nil, fmt.Errorf("missing size for tag name")
+	}
+	numBytesName := int(byteOrder.Uint16(encodedTags[:2]))
+	if numBytesName == 0 {
+		return nil, nil, nil, errEmptyTagNameLiteral
+	}
+	encodedTags = encodedTags[2:]
 
-		if len(encodedTags) < 2 {
-			return nil, false, fmt.Errorf("missing size for tag value: index=%d", i)
-		}
+	bytesName := encodedTags[:numBytesName]
+	encodedTags = encodedTags[numBytesName:]
 
-		numBytesValue := int(byteOrder.Uint16(encodedTags[:2]))
-		encodedTags = encodedTags[2:]
+	if len(encodedTags) < 2 {
+		return nil, nil, nil, fmt.Errorf("missing size for tag value")
+	}
 
-		bytesValue := encodedTags[:numBytesValue]
-		encodedTags = encodedTags[numBytesValue:]
+	numBytesValue := int(byteOrder.Uint16(encodedTags[:2]))
+	encodedTags = encodedTags[2:]
```

Comment:
Just to be sure, please also benchmark the implementation where you would not slide the slice each time, but only use it's original form, and adjust the indices instead.